galsim/compound.py

```diff
@@ -902,3 +905,224 @@ def __setstate__(self, d):
+    read-only properties, are provided as "getters" for the basic parameters:
+
+        .npoints
+        .input_half_light_radius
+        .flux
+        .nstep
+        .gaussians
+            The list of galsim.Gaussian objects representing the points
+        .points
+            The array of x,y offsets used to create the point sources
+
+    Notes
+    -----
+
+    - The algorithm is a modified version of that presented in
+      https://arxiv.org/abs/1312.5514v3 : There is no outer cutoff to how far a
+      point can wander.
+    """
+
+    # these allow use in a galsim configuration context
+
+    _req_params = { "npoints" : int, "half_light_radius" : float }
+    _opt_params = { "flux" : float, "nstep" : int }
+    _single_params = []
+    _takes_rng = True
+
+    # use to convert the requested hlr to a scale factor for each
+    # random walk step
+
+    _interp_ref_fac = 2.09
+    _interp_npts = np.array([6,7,8,9,10,15,20,30,50,75,100,150,200,500,1000])
+    _interp_hlr  = np.array([7.511,7.597,7.647,7.68,7.727,7.827,7.884,7.936,7.974,8.0,8.015,8.019,8.031,8.027,8.043])/8.0
+    _interp_std = np.array([2.043,2.029,1.828,1.817,1.67,1.443,1.235,1.017,0.8046,0.6628,0.5727,0.4703,0.4047,0.255,0.1851])/8.0
+
+    def __init__(self, npoints, half_light_radius, flux=1.0, nstep=40, rng=None, gsparams=None):
+
+        self._half_light_radius = float(half_light_radius)
+
+        self._flux    = float(flux)
+        self._npoints = int(npoints)
+        self._nstep   = int(nstep)
+
+        self._gaussian_sigma = 1.0e-8
+
+        self._input_gsparams=gsparams
+
+        if rng is None:
+            rng = galsim.UniformDeviate()
+
+        self._rng=rng
+
+        self._verify()
+
+
+        factor = self._get_hlr_factor()
+        self._scale = half_light_radius*factor
+
+        self._points = self._get_points()
+        self._gaussians = self._get_gaussians(self._points)
+
+        super(RandomWalk, self).__init__(self._gaussians)
+
+    @property
+    def input_half_light_radius(self):
+        return self._half_light_radius
+
+    @property
+    def flux(self):
+        return self._flux
+
+    @property
+    def npoints(self):
+        return self._npoints
+
+    @property
+    def nstep(self):
+        return self._nstep
+
+    @property
+    def gaussians(self):
+        return self._gaussians
+
+    @property
+    def points(self):
+        return self._points.copy()
+
+    def _get_gaussians(self, points):
+
+        fluxper=self._flux/self._npoints
+        gaussians=[]
+
+        for i in xrange(points.shape[0]):
+            dx,dy = points[i]
+
+            g=galsim.Gaussian(
+                sigma=self._gaussian_sigma,
+                flux=fluxper,
+                gsparams=self._input_gsparams,
+            )
+            g = g.shift(dx=dx, dy=dy)
+
+            gaussians.append(g)
+
+        return gaussians
+
+    def _get_points(self):
+
+        scale=self._scale
+        npoints=self._npoints
+        nstep=self._nstep
+
```

Comment:
There are a number of speed-ups possible in this function, even keeping the current functionality of doing nstep random walk steps, rather than switching to the nstep->inf limit.

First, we can pre-generate a bunch of random values by applying `DeviateNoise` to an image, rather than call `rng()` repeatedly, which suffers from some Python wrapping overhead.
```
rand_values = np.zeros((npoints * nstep,2))
galsim.DeviateNoise(rng).applyTo(galsim.Image(rand_values))
```
Then we can use numpy vectorization to do all the trig and multiplications
```
r_values = scale * rand_values[:,0]
angle_values = 2*np.pi*rand_values[:,1]

dx = r_values * np.cos(angle_values)
dy = r_values * np.sin(angle_values)
```
And finally, it's a bit more efficient to let numpy do the partial sums for you
```
return np.array([ (np.sum(dx[i*nstep : (i+1)*nstep]),
                   np.sum(dy[i*nstep : (i+1)*nstep])) for i in range(npoints) ])
```